lib/web/recordingplayback.go

```diff
@@ -0,0 +1,640 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package web
+
+import (
+	"cmp"
+	"context"
+	"encoding/binary"
+	"fmt"
+	"log/slog"
+	"net/http"
+	"slices"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/gorilla/websocket"
+	"github.com/gravitational/trace"
+	"github.com/hinshun/vt10x"
+	"github.com/julienschmidt/httprouter"
+
+	"github.com/gravitational/teleport"
+	"github.com/gravitational/teleport/api/metadata"
+	apievents "github.com/gravitational/teleport/api/types/events"
+	"github.com/gravitational/teleport/lib/events"
+	"github.com/gravitational/teleport/lib/reversetunnelclient"
+	"github.com/gravitational/teleport/lib/session"
+	"github.com/gravitational/teleport/lib/terminal"
+	"github.com/gravitational/teleport/lib/utils"
+)
+
+// Identifies requests coming from the client (web UI)
+const (
+	// requestTypeFetch requests event data
+	requestTypeFetch byte = 1
+	// requestTypeClose closes the connection
+	requestTypeClose byte = 2
+)
+
+// Response types send back to the client
+const (
+	// eventTypeStart indicates stream start
+	eventTypeStart byte = 1
+	// eventTypeStop indicates stream stop
+	eventTypeStop byte = 2
+	// eventTypeError indicates an error
+	eventTypeError byte = 3
+	// eventTypeSessionStart indicates session started
+	eventTypeSessionStart byte = 4
+	// eventTypeSessionPrint contains terminal output
+	eventTypeSessionPrint byte = 5
+	// eventTypeSessionEnd indicates session ended
+	eventTypeSessionEnd byte = 6
+	// eventTypeResize indicates terminal resize
+	eventTypeResize byte = 7
+	// eventTypeScreen contains terminal screen state
+	eventTypeScreen byte = 8
+	// eventTypeBatch indicates a batch of events
+	eventTypeBatch byte = 9
+)
+
+const (
+	// maxRequestRange is the maximum allowed time range for a request
+	maxRequestRange = 10 * time.Minute
+	// requestHeaderSize is the size of the request header (event type, start time, end time, request ID, and current screen flag)
+	requestHeaderSize = 22
+	// responseHeaderSize is the size of the response header (event type, timestamp, data size, and request ID)
+	responseHeaderSize = 17
+)
+
+// recordingPlayback manages session event streaming
+type recordingPlayback struct {
+	ctx              context.Context
+	cancel           context.CancelFunc
+	clt              events.SessionStreamer
+	sessionID        string
+	logger           *slog.Logger
+	mu               sync.Mutex
+	cancelActiveTask context.CancelFunc
+
+	stream struct {
+		sync.Mutex
+		eventsChan     <-chan apievents.AuditEvent
+		errorsChan     <-chan error
+		lastEndTime    int64
+		bufferedEvents []apievents.AuditEvent
+	}
+
+	websocket struct {
+		sync.Mutex
+		*websocket.Conn
+	}
+
+	terminal struct {
+		sync.RWMutex
```

Comment:
Tim recently showed me that `sync.RWMutex` is a trap and it's usually not an improvement to performance: https://zephyrtronium.github.io/articles/rwmutex.html. So if it's performance you're after here, I'd rethink it.

Reply:
Changed it to a mutex